network/benchmark/src/lib.rs

```diff
@@ -0,0 +1,607 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_config::{
+    config::NodeConfig,
+    network_id::{NetworkId, PeerNetworkId},
+};
+use aptos_logger::{info, warn};
+use aptos_metrics_core::{register_int_counter_vec, IntCounter, IntCounterVec};
+use aptos_network::{
+    application::interface::{NetworkClient, NetworkClientInterface, NetworkServiceEvents},
+    protocols::{network::Event, rpc::error::RpcError, wire::handshake::v1::ProtocolId},
+};
+use aptos_time_service::{TimeService, TimeServiceTrait};
+use aptos_types::{account_address::AccountAddress, PeerId};
+use bytes::Bytes;
+use futures::{
+    channel::oneshot::Sender,
+    stream::{FuturesUnordered, StreamExt},
+};
+use once_cell::sync::Lazy;
+use rand::{rngs::OsRng, Rng};
+use serde::{Deserialize, Serialize};
+use std::{ops::DerefMut, sync::Arc, time::Duration};
+use tokio::{runtime::Handle, select, sync::RwLock};
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[allow(clippy::large_enum_variant)]
+pub enum BenchmarkMessage {
+    DataSend(BenchmarkDataSend),
+    DataReply(BenchmarkDataReply),
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct BenchmarkDataSend {
+    pub request_counter: u64, // A monotonically increasing counter to verify responses
+    pub send_micros: i64, // micro seconds since some epoch at a moment just before this message is sent
+    pub data: Vec<u8>,    // A vector of bytes to send in the request; zero length in reply
```

Comment:
Is my understanding correct that we are BCS encoding the send data? I'm surprised we're getting high performance without using `#[serde(with = "serde_bytes")]` (e.g., [here](https://github.com/aptos-labs/aptos-core/blob/eef00c64113f78e585ce62bfeeaaaca30fd6fad4/consensus/src/dag/types.rs#L605C5-L605C5)). In a previous experiment we saw that the time to serde was fairly large without this.

Reply:
It appears to do the right thing? metrics counters show the expected amount of bandwidth, if I send 1000/s of 100,000 byte messages I get 100MB/s.

Reply:
Yeah, that's weird. I'm fine keeping it this way for now, and I'll play with it when we get a forge job up.